### PR TITLE
Type check with pytype

### LINKED ISSUE #561: Add Typing to Stable Baselines
I'm wondering if it's worth us starting to use Python [type annotations](https://docs.python.org/3/library/typing.html) as in [PEP 484](https://www.python.org/dev/peps/pep-0484/) et al? We can introduce these gradually, and type checkers can still catch some classes of bug even without annotations.

Right now we already specify the types of most functions in the docstring, so I don't think it would be much extra work to specify these formally using type hints.

Concretely we'd need to add type checking to Travis, document it and make sure the current code type checks. I tried running `pytype` against Stable Baselines quickly and there were a few errors (I think mostly false positives) but probably wouldn't take long to fix or suppress them.

@hill-a @araffin @erniejunior @Miffyli curious on your thoughts on this, if there's demand I can probably make a PR for this (though may take me a while).

### PR DESCRIPTION
**Pull request by AdamGleave**
## Description

  - Add type checking with `pytype` to the Travis CI builds; update Dockerfile accordingly.
  - Add `py.typed` file to declare PEP 561 compliance.
  - Fix a few minor bugs caught by `pytype`: mostly wrong variable names.
  - Suppress some false-positives or adapt the code so that it can be type checked.

Note what I have *not* done is added any type annotations (except where it was needed to make code type check). I don't have the bandwidth to do this right now, although it'd be great if we added annotations to some of the core classes.

## Motivation and Context

Closes #561.
- [X] I have raised an issue to propose this change ([required](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [X] I've read the [CONTRIBUTION](https://github.com/hill-a/stable-baselines/blob/master/CONTRIBUTING.md) guide (**required**)
- [X] I have updated the [changelog](https://github.com/hill-a/stable-baselines/blob/master/docs/misc/changelog.rst) accordingly (**required**).
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by AdamGleave** [comment]
@araffin can you rebuild the Docker image to include `pytype`? (Or give me access to DockerHub so I can push a new image.)

**Comment by araffin** [comment]
@AdamGleave sorry for that. I'll try to do that (not sure if i can with the free version)
in the mean time, you can change the travis config to point to your image

**Comment by araffin** [comment]
@AdamGleave what is your docker username?

**Comment by AdamGleave** [comment]
> @AdamGleave what is your docker username?

I'm [https://hub.docker.com/u/adamgleave](adamgleave). Thanks!

**Comment by AdamGleave** [comment]
Travis is now passing. I'm now done with this PR (except any review comments).

One thing we shouldn't forgot is to change `.travis.yml` to point to `araffin/stable-baselines-cpu` since I changed it to point to my account temporarily.

Codacy has some false positives for unused imports that are used in type hints in comments. I think we'll just have to ignore these: it'd probably understand if I used variable type annotations, but that's only supported in 3.6+.

**Comment by araffin** [comment]
>but that's only supported in 3.6+.

Maybe a good time to drop 3.5 support? This would allow to use f-strings and faster dict implementation that keeps the insertion order. I have to check when is the official end of support of 3.5...

> I think we'll just have to ignore these

yes, otherwise, it is not a big deal to ignore those (you have now the rights to do that normally ;) )

**Comment by Miffyli** [comment]
Regarding Python 3.5: Many older linux distros come with Python 3.5 by default and it is difficult-ish to upgrade the pre-packaged one to newer one, so that may be bit of a headache for some (E.g. They need the older distribution/packages to run some specific software). Then again, they could just install minoconda and go with that, so I would vote for moving to Python versions +3.6.

**Comment by AdamGleave** [comment]
> Maybe a good time to drop 3.5 support? This would allow to use f-strings and faster dict implementation that keeps the insertion order. I have to check when is the official end of support of 3.5...
[End of life](https://devguide.python.org/#branchstatus) for 3.5 is 2020-09-13, although I'd be OK dropping support earlier.

File: stable_baselines/common/cmd_util.py [hunk 1]
```diff
@@ -5,11 +5,11 @@
 import os
 
 import gym
-from gym.wrappers import FlattenDictWrapper
+from gym.wrappers import FilterObservation, FlattenObservation
```
Comment (araffin) [inline stable_baselines/common/cmd_util.py]: maybe a try/except for old gym version?

File: docs/misc/changelog.rst [hunk 2]
```diff
@@ -19,6 +19,8 @@ New Features:
 - Add `n_cpu_tf_sess` to model constructor to choose the number of threads used by Tensorflow
 - `VecNormalize` now supports being pickled and unpickled.
 - Add parameter `exploration_initial_eps` to DQN. (@jdossgollin)
+- Add type checking and PEP 561 compliance.
+  Note: most functions are still not annotated, this will be a gradual process.
```
Comment (araffin) [inline docs/misc/changelog.rst]: Missing: the fix for newer gym version

File: stable_baselines/deepq/dqn.py [hunk 3]
```diff
@@ -243,13 +244,16 @@ def learn(self, total_timesteps, callback=None, log_interval=100, tb_log_name="D
                 if can_sample and self.num_timesteps > self.learning_starts \
                     and self.num_timesteps % self.train_freq == 0:
                     # Minimize the error in Bellman's equation on a batch sampled from replay buffer.
+                    # pytype:disable=bad-unpacking
                     if self.prioritized_replay:
+                        assert self.beta_schedule is not None
```
Comment (araffin) [inline stable_baselines/deepq/dqn.py]: error message missing

File: stable_baselines/deepq/experiments/run_atari.py [hunk 4]
```diff
@@ -18,8 +18,6 @@ def main():
     parser.add_argument('--dueling', type=int, default=1)
     parser.add_argument('--prioritized-replay-alpha', type=float, default=0.6)
     parser.add_argument('--num-timesteps', type=int, default=int(10e6))
-    parser.add_argument('--checkpoint-freq', type=int, default=10000)
```
Comment (araffin) [inline stable_baselines/deepq/experiments/run_atari.py]: nice catch, at some we should get rid of those files anyway now that we have the zoo

File: .travis.yml [hunk 5]
```diff
@@ -4,7 +4,7 @@ python:
 
 env:
   global:
-    - DOCKER_IMAGE=araffin/stable-baselines-cpu:v2.7.1
+    - DOCKER_IMAGE=adamgleave/stable-baselines-cpu:v2.9.0
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply .travis.yml]: I've made a `stablebaselines` org and added you as an owner.

File: .gitignore [hunk 6]
```diff
@@ -19,13 +20,11 @@ keys/
 
 # Virtualenv
 /env
-
+/venv
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply .gitignore]: It's duplicated (also present on line 9)

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+COPY ./setup.py /root/code/setup.py
+RUN \
+    mkdir -p ${CODE_DIR}/stable_baselines && \
+    pip install virtualenv && \
+    virtualenv $VENV --python=python3 && \
+    . $VENV/bin/activate && \
+    cd $CODE_DIR && \
+    pip install --upgrade pip && \
+    if [[ $USE_GPU == "True" ]]; then \
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply Dockerfile]: thanks

File: scripts/build_docker.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
+
+CPU_PARENT=ubuntu:16.04
+GPU_PARENT=nvidia/cuda:9.0-cudnn7-runtime-ubuntu16.04
+
+TAG=araffin/stable-baselines
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply scripts/build_docker.sh]: changed to stable baselines

File: scripts/run_docker_cpu.sh [hunk 9]
```diff
@@ -6,7 +6,6 @@ cmd_line="$@"
 echo "Executing in the docker (cpu image):"
 echo $cmd_line
 
-
 docker run -it --rm --network host --ipc=host \
- --mount src=$(pwd),target=/root/code/stable-baselines,type=bind araffin/stable-baselines-cpu:v2.7.1 \
+ --mount src=$(pwd),target=/root/code/stable-baselines,type=bind araffin/stable-baselines-cpu:v2.9.0 \
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply scripts/run_docker_cpu.sh]: changed to stable baselines

File: stable_baselines/common/bit_flipping_env.py [hunk 10]
```diff
@@ -98,7 +98,10 @@ def step(self, action):
         done = done or self.current_step >= self.max_steps
         return obs, reward, done, info
 
-    def compute_reward(self, achieved_goal, desired_goal, _info):
+    def compute_reward(self,
+                       achieved_goal: np.ndarray,
+                       desired_goal: np.ndarray,
+                       _info) -> int:
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply stable_baselines/common/bit_flipping_env.py]: OK, I've updated the constant values to be floats not ints for consistency.

File: .travis.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (araffin) [reply .travis.yml]: good initiative

File: stable_baselines/common/cmd_util.py (hunk 1 above)
Comment (AdamGleave) [reply stable_baselines/common/cmd_util.py]: added

File: docs/misc/changelog.rst (hunk 2 above)
Comment (AdamGleave) [reply docs/misc/changelog.rst]: added

File: stable_baselines/deepq/dqn.py (hunk 3 above)
Comment (AdamGleave) [reply stable_baselines/deepq/dqn.py]: added

**Comment by AdamGleave** [comment]
@araffin thanks for the review, I've addressed the comments.

**Comment by AdamGleave** [comment]
While we're making some changes to the CI, wondering if we should disable Travis branch builds, and just build PRs? Right now our build queue in Travis is often quite long.

We'd need to change `scripts/run_tests_travis.sh` since right now the Codacy upload only happens on branch builds. But this is just becuase `CODACY_PROJECT_TOKEN` may not be available on forks. We could instead directly check for the presence of that environment variable.

**Comment by araffin** [comment]
I agree on both ;)

**Comment by AdamGleave** [comment]
> Almost there, still some edit needed in the doc: https://stable-baselines.readthedocs.io/en/master/guide/install.html#build-the-docker-images

Sorry for missing, updated.

> + maybe a line about running type check?

I've added this in `CONTRIBUTING.md`, couldn't see a better place for it.

> EDIT: I ignored the issues in codacy related to typing

Thanks :)

> EDIT: I'm afraid only @hill-a has the right to change that (for the build)

Travis seems to have let me disable branch builds so I've done that for now (and updated `scripts/run_tests_travis.sh` accordingly).

**Comment by AdamGleave** [comment]
We seem to have both a `travis-ci.com` and `travis-ci.org` instance! The `travis-ci.com` is the one producing "Travis CI - Pull Request " and `travis-ci.org` is the one with "continuous-integration/travis-ci/pr" and "continuous-integration/travis-ci/push".

`travis-ci.org` is deprecated so it's probably best if we just get rid of that one. I've disabled builds on it but can't seem to delete it. @hill-a or someone else with admin access to this repo on GitHub should follow [this](https://help.github.com/en/github/administering-a-repository/enabling-required-status-checks) to remove the status checks beginning with "continuous-integration/"

**Comment by araffin** [comment]
Unfortunately, i think only @hill-a  has the rights to do that. Will ping him again if he does not see this message ;)

**Pull request merged**